std/numeric.d

```diff
@@ -3329,6 +3329,118 @@ C swapRealImag(C)(C input)
     return C(input.im, input.re);
 }
 
+/**
+Finds the mean (colloquially known as the average) of a range. If `r` does
+not provide the `length` member, then this function will do element by
+element summation, rather than the more accurate methods provided by
+$(REF sum, std,algorithm,iteration).
+
+An optional parameter `seed` is used for non-builtin types in order to
+correctly seed the summation with the equivalent to `0`.
+
+This function will return `real.nan` if the range is empty and if the element
+type of `r` is a built-in numerical type. However, if the seed value is defined,
+then empty ranges will return `ElementType!R.init`.
+
+This function is $(BIGOH r.length).
+
+Params:
+    r = An $(REF_ALTTEXT input range, isInputRange, std,range,primitives)
+    seed = An optional parameter for taking the mean of types which can't
+    implicitly convert to `real`. This value will initialize the summation,
+    so it should be equivalent to `0`.
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+    if (isInputRange!R &&
+        is(ElementType!R : real) &&
+        !isInfinite!R)
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    return mean(r, real(0.0));
```

Comment:
Are you aware that `real` FP (in this summation) is _quite_ slower than `double`?

```d
/**
Compile with:
> ldc -release -O3 -mcpu=native -boundscheck=off
*/
import std.algorithm, std.range;

private void doNotOptimizeAway(T)(auto ref T t)
{
    import core.thread : getpid;
    import std.stdio : writeln;
    if(getpid() == 1) {
        writeln(*cast(char*)&t);
    }
}

/// ditto
auto mean(R, T)(R r, T seed)
{
    import std.algorithm.iteration : sum, reduce;
    static if (hasLength!R)
    {
        if (r.length == 1)
            return r.front;

        return sum(r, seed) / r.length;
    }
    else
    {
        import std.typecons : tuple;

        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
            (tuple(size_t(0), seed), r);

        return pair[1] / pair[0];
    }
}

void main() {
    import std.datetime;
    import std.stdio;
    import std.array;
    import std.conv;
    import std.random;
    auto arr = iota(100_000).array;
    auto bench = benchmark!(
        { doNotOptimizeAway({
            return arr.mean(real(0));
        }());},
        { doNotOptimizeAway({
            return arr.mean(double(0));
        }());},
        { doNotOptimizeAway({
            return arr.filter!(a => a >= 0).mean(real(0));
        }());},
        { doNotOptimizeAway({
            return arr.filter!(a => a >= 0).mean(double(0));
        }());},
    )(50_000);
    string[] names = ["mean (real)", "mean (double)", "mean (real, InputRange)", "mean (double, InputRange)"];

    foreach(j,r;bench)
        writefln("%-15s = %s", names[j], r.to!Duration);
}
```

```
> ldc -release -O3 -mcpu=native -boundscheck=off test.d && ./test
mean (real)     = 5 secs, 185 ms, 845 μs, and 3 hnsecs
mean (double)   = 3 secs, 869 ms, 445 μs, and 9 hnsecs
mean (real, InputRange) = 6 secs, 109 ms, 83 μs, and 8 hnsecs
mean (double, InputRange) = 5 secs, 940 ms, 972 μs, and 1 hnsec
```